### ISSUE TITLE
Incorrect tokens when using Replace

### ISSUE THREAD

**Issue by david-waterworth** [huggingface/tokenizers#1043]
This might be related to my other reports on `Replace` but the symptoms are different so I'll create a new issue.

I wanted to remove symbols in my normaliser so I used `normalizers.Replace(tokenizers.Regex(r"\p{S}"), "")`

This is causing results that I don't expect, for example the word `metasys` is in vocab yet I'm seeing

```
tokenizer.encode("metasys").tokens
```
> ['m', 'e', 't', 'a', 's', 'y', 's']

Now what's odd is if I run each step one at a time I get the correct result, i.e.

```
tokenizer.normalizer.normalize_str("metasys")
```
> 'metasys'

'tokenizer.pre_tokenizer.pre_tokenize_str("metasys")'
> [('metasys', (0, 7))]

```
for token in tokenizer.model.tokenize("metasys"):
    print(token.value)
```
> 'metasys'

If I replace `\p{S}` (Symbols) with `\p{P}` (Punctuation) the output of `tokenizer.encode("metasys").tokens` is correct (a single token) but none of the intermediate steps are returning different outputs so I'm a little confused as to what's going on? My original regex was `[\p{P}\p{S}]` - the alternative `[^A-Za-z0-9 ]` also causes the problem but `\p{P}` seems to work so the behaviour seems to be triggered by the regex.

To reproduce I've attached a minimal tokenizers.json
[tokenizer.zip](https://github.com/huggingface/tokenizers/files/9366784/tokenizer.zip)

**Comment by david-waterworth** [comment]
Additionally, I found that using replaces targeted exactly the characters I wanted to eliminate worked i.e.

- Punctuation: `\p{P}` 
- Whitespace other than ' ' [^\S ]
- ASCII 'exotic' non-printables [\x01\x03\x07\x10\x12\x13\x14\x16\x17\x19\x1a\x1d\x1e]

The last one isn't exhaustive, it's only the non-printables I found in my training data (which is - in theory - ascii).

**Comment by Narsil** [comment]
Hi @david-waterworth ,

Do you mind sharing the reproducible tokenizer that triggers this bug ? I wasn't able to reproduce on a test example:

```python
from tokenizers import normalizers, Tokenizer, models, Regex

tokenizer = Tokenizer.from_pretrained("gpt2")
tokenizer.normalizer = normalizers.Replace(tokenizers.Regex(r"\p{S}"), "")

tokenizer.encode("This is a test"),tokens #is, a , test are all in vocabulary
# ['This', 'Ġis', 'Ġa', 'Ġtest']
```

I suspect the problem is that it's a BPE and the merges are not working properly maybe ?

**Comment by david-waterworth** [comment]
It's attached to the original message as a zipped tokeniser.json file that you should be able to load.

**Comment by Narsil** [comment]
Sorry I failed to see it ! 

Ok I have reproduced the bug and can explain how to fix it.

The added tokens are marked as `normalized: true`, what this actually means is that we should normalize the tokens before processing them. with `\p{S}` it means `<s>` becomes `s` which then splits apart `metasys` into `meta + s + y + s` leading to the result you are seeing.

I don't think `normalized` should be the default, but I am not sure, how did you add them ?

Anyway, I could recover the `ids == [5]` and `tokens == ['metasys']` by simply switching to `normalized:false` for the token `<s>` (I would suggest doing it for every token. I don't know you're dataset but usually those special tokens are not present in unnormalized forms in the data, right ?

**Comment by david-waterworth** [comment]
Thanks @Narsil - I'm not actually sure how/when I managed to set `normalized: true` but it's my fault. I've created my own rust script from your unigram trainer where I can train a vocab given a corpus and a manually created seed vocab. Then I manually splice the resulting vocab (without special tokens) into a `tokenizers.json` template. Somehow the special tokens in the template had `normalized: true`
